src/mixin/TransientStorageArray.sol

```diff
@@ -0,0 +1,126 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+pragma solidity ^0.8.28;
+
+/// @title Transient Storage Array
+/// @author CoW DAO developers
+/// @notice Defines an interface to copy a (dynamic-size) bytes array from
+/// calldata to transient storage as well as to read it at a later point in the
+/// same transaction.
+/// @dev The array length is defined as a Solidity transient variable. The
+/// content of the array is stored word-by-word starting from the end of the
+/// transient storage space (slot `type(uint256).max`); each next word is stored
+/// in the transient storage slot immediately before. The endianness of
+/// each word has not been changed.
+/// The array is defined from the end to avoid conflicts with transient
+/// variables that could be relied on by the importing contract. If the contract
+/// that inherits this one only access transient storage through Solidity
+/// variables that are only manipulated through basic Solidity operations (i.e.,
+/// no assembly) then there should be no interferences with this contract.
+/// Otherwise, it's important to have a full understanding of the transient
+/// storage layout of the inheriting contract to avoid that the same transient
+/// slot is overridden when working with two unrelated variables.
+abstract contract TransientStorageArray {
+    uint256 constant BYTES_IN_WORD = 32;
+
+    /// @notice The number of bytes currently stored in the transient storage
+    /// array.
+    uint256 private transient length;
```

Comment:
I assume reading `length` before it was initialized would return 0 which makes it okay to read from the transient storage array before having something written into it?